Item 10230: Entity Grid Action Updates to factor out menu/buttons from create and manage

cnathe opened the pull request:
#### Rationale
The menus on our various App grids have been a source of confusion for users since they don’t use labels that users have in mind for the actions they want to perform when interacting with the application.  The menus also currently group a lot of functionality together that is not necessarily related and, in some cases, we offer options that are superfluous in certain contexts.  With this work, we will reorganize the menus attached to grids to make the actions easier to find and reduce some clutter.

#### Related Pull Requests
* https://github.com/LabKey/labkey-ui-components/pull/810
* https://github.com/LabKey/testAutomation/pull/1090
* https://github.com/LabKey/sampleManagement/pull/924
* https://github.com/LabKey/biologics/pull/1258
* https://github.com/LabKey/inventory/pull/420

#### Changes
* Add new button components: ResponsiveMenuButtonGroup, SamplesAddButton, SamplesAssayButton
* Rename SamplesManageButton to SamplesEditButton
* SubMenu and SubMenuItem changes to support new button submenu cases

cnathe committed:
CreateSamplesSubMenuBase update to add selectedType prop and option to not extract the current menu choice in the submenu case

cnathe committed:
EntityLineageEditMenuItem update to simplify the display text on menu item

cnathe committed:
ManageDropdownButton update to allow for title prop and to remove bsStyle "primary"

cnathe committed:
Rename SamplesManageButton to SamplesEditButton

cnathe committed:
Factor out getMenuItemForSectionKey for use in the single sample derive case

cnathe committed:
Picklist button updates to menu item text and add RequiresPermission check

cnathe committed:
Misc sample action button text updates

cnathe committed:
SubMenu prop to control extractCurrentMenuChoice

cnathe committed:
SubMenuItem update to only render the items if there is no menu text

cnathe committed:
SamplesEditButton - rename from SamplesManageButton, remove picklist related items, remove import, add menu divider

cnathe committed:
v2.154.0-fb-smGridActionsUpdate.0

cnathe committed:
Jest snapshot updates

cnathe committed:
Change RemoveFromPicklistMenuItem to a DisableableButton

cnathe committed:
SampleTabbedGridPanel update for handling the create samples from parent case

cnathe committed:
Jest test fixes

cnathe committed:
Jest test fixes

cnathe committed:
v2.154.0-fb-smGridActionsUpdate.1

cnathe committed:
CreateSamplesSubMenuBase no longer uses QueryGridModel

cnathe committed:
v2.154.0-fb-smGridActionsUpdate.2

cnathe committed:
PicklistButton update to support rendering items as submenu

cnathe committed:
v2.154.0-fb-smGridActionsUpdate.3

cnathe committed:
CreateSamplesSubMenuBase inlineItemsCount prop

cnathe committed:
v2.154.0-fb-smGridActionsUpdate.4

cnathe committed:
SamplesEditButton divider after edit options should only show when edit is shown

cnathe committed:
v2.154.0-fb-smGridActionsUpdate.5

cnathe committed:
Add RequiresPermission check to a few sample grid button cases

cnathe committed:
v2.154.0-fb-smGridActionsUpdate.6

cnathe committed:
Add ResponsiveMenuButtonGroup, SamplesAddButton, SamplesAssayButton, and getSampleWizardURL for use in LKB and LKSM

cnathe committed:
CreateSamplesSubMenuBase - add disabled message for pool samples case

cnathe committed:
v2.154.0-fb-smGridActionsUpdate.7

cnathe committed:
Merge branch 'develop' into fb_smGridActionsUpdate

# Conflicts:
#	packages/components/package.json
#	packages/components/releaseNotes/components.md

cnathe committed:
v2.155.0-fb-smGridActionsUpdate.0

cnathe committed:
Change sample aliquot grid to Delete button instead of Manage button with a single menu item

cnathe committed:
Misc cleanup of unused props

cnathe committed:
Merge branch 'develop' into fb_smGridActionsUpdate

# Conflicts:
#	packages/components/package.json
#	packages/components/releaseNotes/components.md

cnathe committed:
v2.155.1-fb-smGridActionsUpdate.0

cnathe committed:
Remove unused getSelectedSampleTypes()

cnathe committed:
Jest fixes for SamplesEditButton changes

cnathe committed:
yarn run lint-branch-fix

cnathe committed:
v2.155.1-fb-smGridActionsUpdate.1

cnathe committed:
Merge branch 'develop' into fb_smGridActionsUpdate

# Conflicts:
#	packages/components/package.json
#	packages/components/releaseNotes/components.md

cnathe committed:
v2.156.0-fb-smGridActionsUpdate.0

cnathe committed:
update release notes

cnathe committed:
New jest tests

labkey-susanh commented on packages/components/src/internal/components/picklist/PicklistButton.tsx:
```diff
@@ -1,35 +1,49 @@
+                    {items}
+                </DropdownButton>
+            )}
+            {asSubMenu && <SubMenuItem text="Picklists">{items}</SubMenuItem>}
```
Do we want to display these two menu items inline?  I know this is consistent with the other items, and we can make this change in a later story, but the designs show the items not as submenus.

cnathe replied:
Yes. I confirmed this with Andrew and the plan is to show the items under the "More" button inline as much as possible. The "Derive from Selected" and "Assay" items will still be submenus, but the rest should get inlined with dividers. That said, I don't plan to do that as part of this story because the selenium tests are already passing with this display. We will inline the items as part of the next story.

labkey-susanh commented on packages/components/src/internal/components/samples/SamplesAssayButton.tsx:
```diff
@@ -0,0 +1,43 @@
+export const SamplesAssayButton: FC<Props> = memo(props => {
+    const { model, providerType, asSubMenu } = props;
+
+    if (!isSamplesSchema(model?.schemaQuery)) return null;
```
I wonder when this can happen.

cnathe replied:
I think there might be one or two of the "All Samples" tab cases that hit this case and don't show the Assay menu because of it. I think it is an open issue somewhere or behavior we do want to change/allow, but not as part of this story.

labkey-susanh commented on packages/components/src/internal/components/samples/SamplesEditButton.tsx:
```diff
@@ -76,35 +77,15 @@ export const SamplesManageButton: FC<OwnProps & SampleGridButtonProps & Requires
                         <SelectionMenuItem
                             id="update-samples-menu-item"
-                            text="Edit Selected Samples in Grid"
+                            text="Edit in Grid"
```
Should there be a check for `user.canUpdate` with this item?

cnathe replied:
I believe it is not there because there are multiple tabs within the "Edit in Grid" and a user might have perm to edit the storage data without being able to edit the sample data. So I believe that component will do the proper checks.

labkey-susanh commented on packages/components/src/internal/components/samples/models.ts:
```diff
@@ -161,10 +162,8 @@ export interface SampleGridButtonProps {
-    hideButtons?: SamplesManageButtonSections[];
-    excludeStartJob?: boolean;
+    excludedManageMenuKeys?: SamplesEditButtonSections[];
+    hideButtons?: SamplesEditButtonSections[];
```
```suggestion
    hiddenButtons?: SamplesEditButtonSections[];
```
I keep expecting `hideButtons` to be a boolean.

cnathe replied:
Done, changed to hiddenButtons

cnathe committed:
Merge branch 'develop' into fb_smGridActionsUpdate

# Conflicts:
#	packages/components/package.json
#	packages/components/releaseNotes/components.md
#	packages/components/src/internal/components/samples/CreateSamplesSubMenuBase.tsx
#	packages/components/src/internal/components/samples/SamplesTabbedGridPanel.tsx

cnathe committed:
v2.157.0-fb-smGridActionsUpdate.0

cnathe committed:
hideButtons prop renamed to hiddenButtons

cnathe committed:
yarn run lint-branch-fix

cnathe committed:
v2.157.0-fb-smGridActionsUpdate.1

cnathe committed:
Rename hiddenButtons->excludedMenuKeys and rename shouldShowButtons->shouldIncludeMenuItem

cnathe committed:
v2.157.0-fb-smGridActionsUpdate.2

cnathe committed:
v2.157.0-fb-smGridActionsUpdate.3

cnathe committed:
ResponsiveMenuButtonGroup styling to make non-primary buttons look like menus
- border on hover but not initially

cnathe committed:
v2.157.0-fb-smGridActionsUpdate.4

cnathe committed:
Picklist grid edit scroll issue fix
- remove override of css for panel border, leave it match other sample tabbed grid scenarios

cnathe committed:
v2.157.0-fb-smGridActionsUpdate.5

cnathe committed:
yarn run lint-branch-fix

cnathe committed:
update release notes date and version

cnathe committed:
v2.158.0
